ambry-replication/src/main/java/com.github.ambry.replication/ReplicationEngine.java

```diff
@@ -262,105 +221,106 @@ public void shutdown() throws ReplicationException {
+      remoteReplicaInfo.getReplicaThread().removeRemoteReplicaInfo(remoteReplicaInfo);
+      remoteReplicaInfo.setReplicaThread(null);
     }
-    dataNodeRemoteReplicaInfosPerDC.put(datacenter, dataNodeRemoteReplicaInfos);
   }
 
   /**
-   * Partitions the list of data nodes between given set of replica threads for the given DC
+   * Assign {@link RemoteReplicaInfo} to a {@link ReplicaThread} for replication.
+   * The assignment is based on {@link DataNodeId}. If no {@link ReplicaThread} responsible for a {@link DataNodeId},
+   * a {@link ReplicaThread} will be selected by {@link ReplicationEngine#getReplicaThreadIndexToUse(String)}.
+   * Create threads pool for a DC if not exists.
+   * @param remoteReplicaInfos List of {@link RemoteReplicaInfo} to add.
+   * @param startThread if threads need to be started when create.
    */
-  private void assignReplicasToThreadPool() throws IOException {
-    for (Map.Entry<String, DataNodeRemoteReplicaInfos> mapEntry : dataNodeRemoteReplicaInfosPerDC.entrySet()) {
-      String datacenter = mapEntry.getKey();
-      DataNodeRemoteReplicaInfos dataNodeRemoteReplicaInfos = mapEntry.getValue();
-      Set<DataNodeId> dataNodesToReplicate = dataNodeRemoteReplicaInfos.getDataNodeIds();
-      int dataNodesCount = dataNodesToReplicate.size();
-      int replicaThreadCount = numberOfReplicaThreads.get(datacenter);
-      if (replicaThreadCount <= 0) {
-        logger.warn("Number of replica threads is smaller or equal to 0, not starting any replica threads for {} ",
-            datacenter);
-        continue;
-      } else if (dataNodesCount == 0) {
-        logger.warn("Number of nodes to replicate from is 0, not starting any replica threads for {} ", datacenter);
-        continue;
-      }
-
-      // Divide the nodes between the replica threads if the number of replica threads is less than or equal to the
-      // number of nodes. Otherwise, assign one thread to one node.
-      logger.info("Number of replica threads to replicate from {}: {}", datacenter, replicaThreadCount);
-      logger.info("Number of dataNodes to replicate :", dataNodesCount);
-
-      if (dataNodesCount < replicaThreadCount) {
-        logger.warn("Number of replica threads: {} is more than the number of nodes to replicate from: {}",
-            replicaThreadCount, dataNodesCount);
-        replicaThreadCount = dataNodesCount;
-      }
-
-      ResponseHandler responseHandler = new ResponseHandler(clusterMap);
+  protected void addRemoteReplicaInfoToReplicaThread(List<RemoteReplicaInfo> remoteReplicaInfos, boolean startThread) {
+    for (RemoteReplicaInfo remoteReplicaInfo : remoteReplicaInfos) {
+      DataNodeId dataNodeIdToReplicate = remoteReplicaInfo.getReplicaId().getDataNodeId();
+      String datacenter = dataNodeIdToReplicate.getDatacenterName();
+      createThreadPoolIfNecessary(datacenter, startThread);
+      ReplicaThread replicaThread = dataNodeIdToReplicaThread.computeIfAbsent(dataNodeIdToReplicate,
```

Comment:
Minor: throw IllegalStateException if threadpool was not created for datacenter, to avoid NPE.